rfd/9999-webauthn-support.md

```diff
@@ -0,0 +1,501 @@
+is implemented via the
+[navigator.credentials](https://developer.mozilla.org/en-US/docs/Web/API/CredentialsContainer)
+object, which is WebAuthn-exclusive and
+[distinct from the U2F APIs in use](https://github.com/google/u2f-ref-code#reference-code-for-u2f-specifications).
+
+WebAuthn is backwards compatible with U2F in the sense that FIDO/CTAP1 devices,
+ie U2F-capable devices, are still supported. The main difference in
+implementations lies in the definition of the App ID / Relying Party ID (ie,
+"https://example.com:3080" vs "example.com"). The "appid" extension is the
+solution for this problem and seems to be well-supported for browsers
+(__TODO(codingllama):__ Verify for Firefox and Safari).
+
+Unfortunately, the appid extension is not without its issues.
+[As the standard itself says in item 2](https://www.w3.org/TR/webauthn/#sctn-appid-extension),
+_"When verifying the assertion, expect that the rpIdHash MAY be the hash of the
+AppID instead of the RP ID."_ This advice
+[wasn't followed by github.com/duo-labs/webauthn](https://github.com/duo-labs/webauthn/blob/9f1b88ef44cc0e4f5ddf511ed12a3aa468f972d7/protocol/assertion.go#L117),
+but in the short-term may be circumvented by creating a secondary
+[WebAuthn object](https://pkg.go.dev/github.com/duo-labs/webauthn/webauthn#WebAuthn)
+with RFID = AppID, used exclusively to validate legacy requests (identified by
+the presence of the "incorrect" rpIdHash).
+
+### UX and configuration
+
+WebAuthn support is designed to be as seamless as the U2F experience. Users
+enabling MFA for the first time may register their devices via `tsh mfa add` and
+use `tsh login` as usual. Users migrating from U2F to WebAuthn retain their
+previously-registered devices and keep their workflow unchanged.
+
+Migration from U2F to WebAuthn is achieved simply by changing the second_factor
+option from "u2f" to "webauthn". The WebAuthn backend sends challenges for all
+previously-registered U2F devices, in addition to any WebAuthn-registered
+devices, so migration is a seamless process - no re-registration required.
+
+If the second factor setting is "on", then the backend sends both U2F and
+WebAuthn challenges. In this case user-facing interfaces, such as `tsh` and the
+Proxy UI, should favor replying to WebAuthn challenges (instead of U2F).
+
+Additions to `teleport.yaml` are as follows:
+
+```yaml
+auth_service:
+  authentication:
+    type: local
+    second_factor: webauthn # "on" allows all MFA options
+    webauthn:
+      relying_party:
+        display_name: "Teleport Auth"
+        # ID of the Relying Party, ie, the Teleport service performing user
+        # authentication.
+        # Must be a domain name, such as "example.com".
+        # Defaults to the Proxy domain name.
+        # IMPORTANT: similarly to the U2F app ID, if the Relying Party ID
+        # changes then all devices must be re-registered.
+        id: localhost
```

Comment:
Considering how both CLI and Proxy UI talk to the Proxy for login, it's difficult to see this being different (technically you could use the Auth server both for ID and Origin, but that would fail all our current tools/interfaces).

It's important for WebAuthn, especially when browser-based, that both RPID and Origin match the backend returning/verifying the challenges. The implications of this setting are that authn goes through the Proxy, which sounds correct to me.

Simplified configuration to omit RPID.